bundler/lib/bundler/errors.rb

```diff
@@ -172,4 +172,19 @@ def initialize(orig_exception, msg)
 
     status_code(36)
   end
+
+  class InsecureInstallPathError < BundlerError
+    def initialize(path)
+      @path = path
+    end
+
+    def message
+      "Your installation path #{@path} has insecure permissions (it's both " \
+      "world-writable and does not have the sticky bit set), so Bundler can't " \
+      "reinstall any gems into it since it needs to remove previous installations " \
+      "which is insecure."
```

Comment:
This is a bit of a run on sentence. I had to dig into this a little to understand the vulnerability. Let me see if I can break it down.

- Installation path is insecure.
- It's insecure because it's world-writeable without sticky bit.
- World writeable would be ok with a sticky bit, but without it we're refusing to write to it since it's globally writeable by any user and that's unsafe for rubygems.
- Bundler is refusing to write to that directory since it exposes the user to [TOCTTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) as detailed in the FileUtils#remove_entry_secure docs.
- Since we can't remove the old files, we refuse to install.

There was a previous message that was removed. If I tried to combine the two so that it explained what I just learned by digging in, I'd offer something like:

```ruby
      "The installation path is insecure. Bundler cannot continue.\n" \
      "#{@path} is world-writable without a sticky bit.\n"
      "FileUtils#remove_entry_secure prevents removing these files to protect against a time-based vulnerability. Please change the permissions of this directory or choose a different install path."
```

You can of course format or change however you wish. Just trying to find a language that would have made sense to me initially.

Reply:
Thanks for digging in!

I agree your message is better, but I'd have two nits:

* Is mentioning `FileUtils.remove_entry_secure` too much of an implementation detail to mention here? Should we say `Bundler` instead?
* Should we mention why we are trying to delete a directory? Maybe it will not be straightforward to some users why `bundle install` deletes things.

Reply:
Yes on both points. So something like "Bundler cannot safely replace gems in world writeable directories due to the potential for time-based vulnerabilities."

Reply:
Yeah, that looks great, I'll update the message plus work on a test!

Reply:
Message updated 👍.